src/canary.rs

```diff
@@ -233,56 +232,165 @@ macro_rules! assert_subroutine {
 mod paint_subroutine {
     use super::*;
 
-    /// Execute the subroutine.
+    /// Write [`CANARY_U32`] to the stack.
     ///
     /// ## Assumptions
     /// - Expects the [`Core`] to be halted and will leave it halted when the function
     /// returns.
     /// - `low_addr` and `size` need to be 4-byte-aligned.
     pub fn execute(core: &mut Core, low_addr: u32, stack_size: u32) -> Result<(), probe_rs::Error> {
         assert_subroutine!(low_addr, stack_size, self::SUBROUTINE.len() as u32);
+        super::execute_subroutine(core, low_addr, stack_size, self::SUBROUTINE)?;
+        self::overwrite_subroutine(core, low_addr)?;
+        Ok(())
+    }
 
-        // prepare subroutine
-        let previous_pc = super::prepare_subroutine(core, low_addr, stack_size, self::SUBROUTINE)?;
-
-        // execute subroutine and wait for it to finish
-        core.run()?;
-        core.wait_for_core_halted(TIMEOUT)?;
-
-        // overwrite subroutine
-        core.write_8(low_addr as u64, &[CANARY_U8; self::SUBROUTINE.len()])?;
-
-        // reset PC to where it was before
-        core.write_core_reg(PC, previous_pc)
+    fn overwrite_subroutine(core: &mut Core, low_addr: u32) -> Result<(), probe_rs::Error> {
+        core.write_8(low_addr as u64, &[CANARY_U8; self::SUBROUTINE.len()])
     }
 
     const SUBROUTINE: [u8; 12] = [
         0x88, 0x42, // cmp      r0, r1
-        0x01, 0xd8, // bhi.n    200f4 <paint+0x8>
+        0x01, 0xd8, // bhi.n    #6 <paint+0x8>
         0x04, 0xc0, // stmia    r0!, {r2}
-        0xfb, 0xe7, // b.n      200ec <paint>
+        0xfb, 0xe7, // b.n      #0xfffffffa <paint>
         0x00, 0xbe, // bkpt     0x0000
         0x00, 0xbe, // bkpt     0x0000 (padding instruction)
     ];
 }
 
-/// Prepare target to execute subroutine.
+/// Measure-stack subroutine.
+///
+/// ### Corresponds to following rust code
 ///
-/// After calling this function, the program counter will be at the beginning of
-/// the subroutine.
+/// ```rust
+/// #[export_name = "measure"]
+/// unsafe fn measure(mut low_addr: u32, high_addr: u32, pattern: u32) -> u32 {
+///     let mut result = 0;
+///
+///     while low_addr < high_addr {
+///         if (low_addr as *const u32).read() != pattern {
+///             result = low_addr;
+///             break;
+///         } else {
+///             low_addr += 4;
+///         }
+///     }
+///
+///     result
+/// }
+/// ```
+///
+/// ### Generated assembly
+///
+/// The assembly is generated from aboves rust code, using the jorge-hack.
+///
+/// ```armasm
+/// 000200ec <measure>:
+///     200ec:    4288    cmp      r0, r1
+///     200ee:    d204    bcs.n    #0xc <measure+0xe>
+///     200f0:    6803    ldr      r3, [r0, #0]
+///     200f2:    4293    cmp      r3, r2
+///     200f4:    d102    bne.n    #8 <measure+0x10>
+///     200f6:    1d00    adds     r0, r0, #4
+///     200f8:    e7f8    b.n      #0xfffffff8 <measure>
```

Comment:
`#0xfffffff8` is `#-8`

Reply:
Nice. I was thinking about that, but wasn't sure if that is "allowed".

Reply:
I think you can do what you like in the comments :)

Reply:
That's true. I just thought I would like to keep it in a format an assembler would accept. So people don't need to guess what I meant, but can just read it up in some documentation.

I just tried, but I couldn't figure out any way to get this to work with `arm-none-eabi-as`, neither with `#0xfffffff8`, nor with `#-8`. So I will just go with `#-8`, because, as you said, it is more clear.